### PR TITLE
AO3-4450 Rspec pseud test expansion

### PR DESCRIPTION
**Pull request by cesy**
Updated summary: This is now just enhancements to the pseud spec and should be mergeable.
https://otwarchive.atlassian.net/browse/AO3-4450

Most of this was written 4 years ago, but I'm putting it out there in case some of it's still useful. Open questions:

* It looks like we don't have any view specs, do we want those?
* Is webrat still the best gem or should we research any new alternatives?
* It beefs up the pseud model spec a little, which is probably useful - I'm happy to split that bit out into a clean PR if people like it
* We can probably scrap that draft in work_spec - I'll tidy up the commits if people want this
* Various bits to make it work, probably need some more commenting as I can't remember why in a few cases

### CODE REVIEW

**Comment by ariana-paris** [comment]
We seem to currently only test the views through Cucumber, so Rspec is used for lower-level unit tests (and probably only covering areas that were developed by people who felt comfortable with it at that). We've tended over the past couple of years to add new features rather than specs for new views, which at least has the benefit of consistency and readability for non-coders (in theory) - even though this also means our test suite is very slow.

As to webrat, a superficial look at their Github pages suggests that Capybara, which we currently use, is in more active development, so I don't know if the webrat features make its adoption worthwhile. There might also be something even more shiny out there!

Beyond that, the fact that the tests are failing on Travis will need to be addressed before this can really be reviewed, and I think some of Hound's comments are worth looking at, notably concerning tab characters which we've been trying to eradicate for years! But the pseud tests look potentially interesting, though it would be worth checking whether Cucumber tests already have that functionality.

**Comment by cesy** [comment]
Okay, I think I will have a look at fixing/de-duping the pseud spec stuff and saving anything useful, then, and scrap the view spec stuff, unless anyone has a different view.

File: spec/models/pseud_spec.rb [hunk 1]
```diff
@@ -16,4 +16,42 @@
...
+  describe "save" do
+
+		before do
+			@user = User.new
+      @user.login = "mynamepseud"
+      @user.age_over_13 = "1"
+      @user.terms_of_service = "1"
+      @user.email = "foo1@archiveofourown.org"
+      @user.password = "password"
+      @user.save
```
Comment (cesy) [inline spec/models/pseud_spec.rb]: I can't figure this one out - I think I've fixed all the tabs at that line.

File: spec/models/pseud_spec.rb [hunk 2]
```diff
@@ -16,4 +16,42 @@
...
+      @pseud.user_id = @user.id
+      @pseud.name = "MyName"
+    end
+    
+    it "should save a minimalistic pseud" do
+      @pseud.should be_valid_verbose
+      @pseud.save.should be_true
+      @pseud.errors.should be_empty
+    end
+
```
Comment (cesy) [inline spec/models/pseud_spec.rb]: Still puzzled how this is inconsistent - it looks sensible to me.

File: spec/support/matchers/be_valid_verbose.rb [hunk 3]
```diff
@@ -0,0 +1,17 @@
+RSpec::Matchers.define :be_valid_verbose do
+  match do |model|
```
Comment (cesy) [inline spec/support/matchers/be_valid_verbose.rb]: I'm not confident of changing this one and it doing the right thing. Can anyone easily explain it, or is it okay as-is?

File: spec/models/pseud_spec.rb [hunk 4]
```diff
@@ -16,4 +16,42 @@
...
\ No newline at end of file
+  describe "save" do
+
+		before do
+			@user = User.new
+      @user.login = "mynamepseud"
+      @user.age_over_13 = "1"
+      @user.terms_of_service = "1"
+      @user.email = "foo1@archiveofourown.org"
+      @user.password = "password"
```
Comment (cesy) [inline spec/models/pseud_spec.rb]: Any ideas what it's seeing that I'm not? This looks consistent to me.

Comment (sarken) [inline spec/models/pseud_spec.rb]: It shows up in my email with before do and @user = User.new indented with tabs, if that helps!

Comment (cesy) [inline spec/models/pseud_spec.rb]: I can see that here, but I've fixed both in the code, and the latest view on Github doesn't show them, so why hasn't Hound's comment disappeared?

File: spec/models/pseud_spec.rb (hunk 1 above)
Comment (ariana-paris) [inline spec/models/pseud_spec.rb]: The tabs are not appearing in Github either now so I guess we'll have to chalk it up to a glitch somewhere

File: spec/support/matchers/be_valid_verbose.rb (hunk 3 above)
Comment (ariana-paris) [inline spec/support/matchers/be_valid_verbose.rb]: I think it means replace that whole block with `match(&:valid?)` - I'm guessing `model` is the only thing inside the `match` object so you can use the `&:` shorthand to call `valid?` on it without any explicit unwrapping. *makes a personal note of the shorthand for future use*

**Comment by scottsds** [comment]
These tests look good to me now. Thanks Cesy!

**Pull request merged**